README.md
# HumanPassword
Simple library to generate a password (string + number) easy to remember.

[![Build Status](https://travis-ci.org/fabioricali/HumanPassword.svg?branch=master)](https://travis-ci.org/fabioricali/HumanPassword) [![Coverage Status](https://coveralls.io/repos/github/fabioricali/HumanPassword/badge.svg?branch=master)](https://coveralls.io/github/fabioricali/HumanPassword?branch=master)

## Installation

```javascript
npm install human-password --save
```

## Example
### Basic
```javascript
var humanPassword = require('human-password');

humanPassword();
// => 'fuxeru9070'
```

### With options

Name | Type | Default | Description
-|-|-|-
couples | integer | 3 | Couple of consonant + vowel
digits | integer | 4 | Number of digits, if is 0 number will be hidden
randomUpper | boolean | false | Random letters uppercase
numberPosition | string | end | Number position in string, can be "start", "middle", "end" and "random"

```javascript
var humanPassword = require('human-password');

humanPassword({
    couples: 5,
    digits: 4,
    randomUpper: true
});
// => 'PutIGoKobi7136'

// Hide number
humanPassword({
    couples: 5,
    digits: 0
});
// => 'gicominobi'

// Number in start position
humanPassword({
    couples: 5,
    digits: 4,
    numberPosition: 'start'
});
// => '6539vikohylure'
```

## License
HumanPassword is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT)

## Author
[Fabio Ricali](http://rica.li)

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.3] - 2017-06-01
### Added
- CHANGELOG.md
- New position "middle" and "random" for numberPosition option

### Changed
- License

## [1.0.2] - 2017-05-29
### Added
- CHANGELOG.md
- New param "numberPosition"

### Changed
- Now if "digits" param equal "0", the number will be hidden
- README.md

## [1.0.0] - 2017-05-28

test/all.js
/**
 * Created by Fabio on 28/05/2017.
 */
const assert = require('assert');
const humanPassword = require('../index');

describe('HumanPassword test', function () {

    describe('generate', function () {

        it('should be return 10', function () {
            let result = humanPassword({
                digits: 4,
                couples: 3
            });
            console.log(result);
            assert.equal(10, result.length);
            assert.equal(true, typeof result.substring(0, 6) === 'string');
            assert.equal(true, typeof parseInt(result.substring(6, 10)) === 'number' && result.substring(6, 10).length === 4);
        });

        it('should be return 10 with number position start', function () {
            let result = humanPassword({
                digits: 4,
                couples: 3,
                numberPosition: 'start'
            });
            console.log(result);
            assert.equal(10, result.length);
            assert.equal(true, typeof parseInt(result.substring(0, 4)) === 'number' && result.substring(0, 4).length === 4);
            assert.equal(true, typeof result.substring(4, 10) === 'string');
        });

        it('should be return 10 with number position middle', function () {
            let result = humanPassword({
                digits: 4,
                couples: 3,
                numberPosition: 'middle'
            });
            console.log(result);
            console.log(result.substring(3, 7));
            console.log(result.substring(7, 10));
            assert.equal(10, result.length);
            assert.equal(true, typeof parseInt(result.substring(3, 7)) === 'number' && result.substring(3, 7).length === 4);
            assert.equal(true, typeof result.substring(7, 10) === 'string');
        });

        it('should be return 10 with number position random', function () {
            let result = humanPassword({
                digits: 4,
                couples: 3,
                numberPosition: 'random'
            });
            console.log(result);
            assert.equal(10, result.length);
        });

        it('should be return 6 with digits = 0', function () {
            let result = humanPassword({
                digits: 0,
                couples: 3
            });
            console.log(result);
            assert.equal(6, result.length);
        });

        it('should be return error with digits as string', function (done) {
            try{
                humanPassword({
                    digits: 'esdsdsd'
                });
            }catch(e) {
                done()
            }
        });

        it('should be return error with couples as string', function (done) {
            try{
                humanPassword({
                    couples: 'hhhhahha'
                });
            }catch(e) {
                done()
            }
        });

        it('should be return error with randomUpper as string', function (done) {
            try{
                humanPassword({
                    randomUpper: 'esdsdsd'
                });
            }catch(e) {
                done()
            }
        });

        it('should be return error with numberPosition as boolean', function (done) {
            try{
                humanPassword({
                    numberPosition: true
                });
            }catch(e) {
                done()
            }
        });

        it('should be return error with numberPosition equal different', function (done) {
            try{
                humanPassword({
                    numberPosition: 'left'
                });
                done('failed');
            }catch(e) {
                done()
            }
        });

    });

    describe('randomNumber', function () {
        it('should be return 4 digits', function () {
            let result = humanPassword._randomNumber(4);
            assert(result.toString().length, 4)
        });

        it('should be return error with string as param', function (done) {
            try {
                humanPassword._randomNumber('4');
            } catch (e) {
                done();
            }
        });
    });

    describe('randomBoolean', function () {
        it('should be return type boolean', function () {
            assert('boolean', typeof humanPassword._randomBoolean());
        });
    });

    describe('randomItem', function () {
        it('should be return type an item', function () {
            assert(1, humanPassword._randomItem([1]));
        });

        it('should be return error', function (done) {
            try {
                humanPassword._randomItem('aa')
            } catch (e) {
                done();
            }
        });
    });

});

index.js
/**
 * Created by Fabio on 28/05/2017.
 */

/**
 * Assets
 * @type {{vowels: [*], consonants: [*]}}
 */
const assets = {
    vowels: ['a', 'e', 'i', 'o', 'u', 'y'],
    consonants: ['b', 'c', 'd', 'f', 'g', 'h', 'j', 'k', 'l', 'm', 'n', 'p', 'q', 'r', 's', 't', 'v', 'w', 'x', 'z']
};

/**
 * Default option
 * @type {{digits: number, couples: number, randomUpper: boolean, numberPosition: string}}
 * @private
 */
let _opt = {
    digits: 4,
    couples: 4,
    randomUpper: false,
    numberPosition: 'end' // start|middle|end|random
};

/**
 * Return a random number
 * @param digits
 * @returns {number}
 */
function randomNumber(digits = 4) {
    if (typeof digits !== 'number')
        throw new Error('digits must be a number');

    let zero = '';
    for (let i = 0; i < digits - 1; i++)
        zero += '0';

    return Math.floor(parseInt(1 + zero) + Math.random() * parseInt(9 + zero));
}

/**
 * Extract a random item from an array
 * @param items
 * @returns {*}
 */
function randomItem(items) {
    if (typeof items !== 'object')
        throw new Error('items must be an array');

    return items[Math.floor(Math.random() * items.length)];
}

/**
 * Return random boolean
 * @returns {boolean}
 */
function randomBoolean() {
    return Math.random() >= 0.5;
}

/**
 * Generate password
 * @param opt
 * @returns {string}
 */
function generate(opt) {

    if (typeof opt === 'object')
        for (let i in opt) {
            if (opt.hasOwnProperty(i)) {
                _opt[i] = opt[i];
            }
        }

    if (typeof _opt.couples !== 'number')
        throw new Error('couples must be a number');

    if (typeof _opt.randomUpper !== 'boolean')
        throw new Error('randomUpper must be a boolean');

    if (typeof _opt.digits !== 'number')
        throw new Error('digits must be a number');

    if (typeof _opt.numberPosition !== 'string')
        throw new Error('numberPosition must be a string ("start" or "end")');

    let str = '', consonant, vowel;

    for (let i = 0; i < _opt.couples; i++) {
        consonant = randomItem(assets.consonants);
        vowel = randomItem(assets.vowels);
        if (_opt.randomUpper) {
            if (randomBoolean())
                consonant = consonant.toUpperCase();

            if (randomBoolean())
                vowel = vowel.toUpperCase();
        }
        str += consonant + vowel;
    }

    if (_opt.digits > 0) {
        if (_opt.numberPosition === 'start')
            str = randomNumber(_opt.digits) + str;
        else if (_opt.numberPosition === 'middle') {
            let startPos = Math.round(str.length / 2);
            let number = randomNumber(_opt.digits).toString();
            str = [str.slice(0, startPos), number, str.slice(startPos)].join('');
        } else if (_opt.numberPosition === 'random') {
            let startPos = Math.floor(Math.random() * (str.length + 1));
            let number = randomNumber(_opt.digits).toString();
            str = [str.slice(0, startPos), number, str.slice(startPos)].join('');
        } else if (_opt.numberPosition === 'end')
            str += randomNumber(_opt.digits);
        else
            throw new Error('numberPosition can be equal to "start" or "end"');
    }

    return str;
}

/**
 * Exports constructor
 * @type {generate}
 */
module.exports = generate;

/**
 * Exports for testing
 * @type {randomNumber}
 * @private
 */
module.exports._randomNumber = randomNumber;

/**
 * Exports for testing
 * @type {randomItem}
 * @private
 */
module.exports._randomItem = randomItem;

/**
 * Exports for testing
 * @type {randomBoolean}
 * @private
 */
module.exports._randomBoolean = randomBoolean;